moveit_core/collision_detection/include/moveit/collision_detection/allvalid/collision_detector_allocator_allvalid.h

```diff
@@ -39,13 +39,23 @@
 #include <moveit/collision_detection/collision_detector_allocator.h>
 #include <moveit/collision_detection/allvalid/collision_env_allvalid.h>
 
+#ifdef ROS_BUILD_SHARED_LIBS // ros is being built around shared libraries
+  #ifdef moveit_collision_detection_EXPORTS // we are building a shared lib/dll
+    #define MOVEIT_COLLISION_DETECTION_DECL ROS_HELPER_EXPORT
+  #else // we are using shared lib/dll
+    #define MOVEIT_COLLISION_DETECTION_DECL ROS_HELPER_IMPORT
+  #endif
+#else // ros is being built around static libraries
+  #define MOVEIT_COLLISION_DETECTION_DECL
+#endif
+
 namespace collision_detection
 {
 /** \brief An allocator for AllValid collision detectors */
 class CollisionDetectorAllocatorAllValid
   : public CollisionDetectorAllocatorTemplate<CollisionEnvAllValid, CollisionDetectorAllocatorAllValid>
 {
 public:
-  static const std::string NAME;  // defined in collision_env_allvalid.cpp
+  MOVEIT_COLLISION_DETECTION_DECL static const std::string NAME;  // defined in collision_env_allvalid.cpp
```

Comment:
how about changing it to `static constexpr char[] NAME { ... };` instead? does that still require the helper?

Reply:
yes, I think it will still need `__dllexport` in front of the static variable for it to be properly exported into the DLL (and usable from other libraries).

Reply:
ttbomk the NAME strings *should not* be used outside the class, they are just used to name `ROS_*` log methods in the class.
So I'm wondering why these are a public interface anyway?

`static constexpr char[] NAME` in the corresponding sources seems like the better approach here (and wouldn't need these cumbersome exports).

Reply:
Note: If we really need the DECL macros, we should provide the helpers in moveit_core/macros and not require pre-processor logic in every single module.

Reply:
I think what you caught here was us having that NAME variable defined in the class when (imo) it should just be constexpr in the source file as it is not part of the external interface and is only for log namespacing as @v4hn said.  I can try to fix these for you and make a small PR against your branch for that.

Reply:
that would be great. I just fixed pre-commit and tried to make our CMake changes backward compatible. let's see what the combined CI's say.

Reply:
I finally got around to looking at this and this `NAME` variable is not what we thought it was @v4hn.  It is not used in ROS_*_LOG messages and is instead part of the public interface of this class.  Each of the `collision_detection::CollisionDetectorAllocator*` classes have a NAME field.  These are used by the getName field and is part of the public interface.  This seems odd to me but it isn't what we thought it was.

I'm not sure what the right approach is.  I know for moveit2 the approach we saw taken was to define these macros in a separate moveit_common package that all the packages could depend on and get the export symbol header from there.

Have you tried the cmake option: `set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)` as that would make it simpler if it worked.    It looks like [from the documentation](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) to use this you'd need cmake 3.4 or newer on windows.  Would that be a possible option?

Reply:
See https://github.com/ros-planning/moveit2/pull/286 for a discussion around `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`

Reply:
@tylerjw In this case I would argue we should refactor the API to a simple virtual getter method instead.
Exposing a static string as API is awkward at least and brings us more trouble than necessary in these efforts.

Reply:
I agree that the static string is awkward and should just be refactored away.

Reply:
This can be resolved.